test/unit/Container.spec.ts

```diff
@@ -128,13 +112,13 @@ describe("Container", function() {
             }
         }
 
-        function createCar (engine: Engine) {
-            return new Car(engine);
+        class CarFactory {
+            @Factory()
+            public static createCar (engine: Engine): Car {
```

Comment:
Actually, I think that concepts of `@Service` and `@Factory` decorators are very similar. The difference is that service is created by calling the constructor of the class directly and factory is used to instantiate another class. Both of them have some "magic" in it. The entire concept of autowiring is about implicit configuration. So, I believe this could be acceptable. In order to mitigate such conflicts we can explicitly check if factory is already registered for a specified service and throw an exception to make such conflicts apparent.

However, I agree that `@Service({ factory: carFactory })` syntax is more expressive and I like it better. Maybe we could have both of them?

The more serious question is I can't really define how the factory itself should look like. Could it be a simple function? Or should it be a static class with static function? Of should it be a normal class, which itself must be created before it's method is called in order to create a service? Probably, we should support all these scenarios? In such case `@Factory` decorator is not an option, because it can't be added to a simple function due to language limitations.

Finally, how do we support all those use cases? I will try to summarize them:

```ts
class Car {
  constructor (private engine: Engine) {
  }
}
```

**1. The factory is just a simple standalone function.**

```ts
@Service({ factory: createCar })
class Car {
  constructor (private engine: Engine) {
  }
}

function createCar (Engine: engine): Car {
  return new Car(engine);
}
```

**2. The factory is a static function of a class.**

```ts
@Service({ factory: CarFactory.createCar })
class Car {
  constructor (private engine: Engine) {
  }
}

class CarFactory {
  public static createCar (Engine: engine): Car {
    return new Car(engine);
  }
}
```

**3. The factory is a normal function of a class instance.**

```ts
@Service({ factory: ... })
class Car {
  constructor (private engine: Engine) {
  }
}

class CarFactory {
  construct (private logger: Logger) {
  }
  public createCar (Engine: engine): Car {
    this.logger.debug("Creating a new car!");
    return new Car(engine);
  }
}
```

---

The first two scenarios looks almost the same, the function is called directly.

However, the third case is interesting, because in order to call a factory function we need to create an instance of a factory class first. So the factory class could be a service in itself. The question is how should we specify both the class of the factory and the method?

`@Service({ factory: [CarFactory, CarFactory.prototype.createCar] }) class Car {}` or

`@Service({ factory: [CarFactory, "createCar"] }) class Car {}` ?

The first variant looks longer, but more specific and could be parsed and understood by IDE.

What do you think about all this?